subscription CE extension spec proposal (#4881)
knative/eventing: docs/spec/sources.md

    // Source.
    // +optional
    SinkURI *apis.URL `json:"sinkUri,omitempty"`

    // CloudEventAttributes are the specific attributes that the Source uses

>     // as part of its CloudEvents.

Reviewer: I'm not sure I understand what you mean by "attributes that the source *uses* as part of its CloudEvents" . Do you mean "generates/emits" ?

Author: Sorry that's just copy&paste from the existing code. The source uses these attributes when emitting events?

Reviewer 2: I does sound ambiguous, but like Lionel said, it's copied from the code directly. Maybe it can be clarified in a separate PR.

Author: marking as resolved as I would like this PR to be only about the new CE attribute.

    // +optional
    CloudEventAttributes []CloudEventAttributes `json:"ceAttributes,omitempty"`
}
```

> For a full definition of `Status`, `SinkURI`, `CloudEventAttributes` and please

Reviewer 2: I think you meant:

```suggestion
For a full definition of `Status`, `SinkURI` and `CloudEventAttributes`, please
```

Author: fixed.

> see [Status](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#Status), and

Reviewer: ```suggestion
see [Status](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1#SourceStatus), and
```

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

> A source that generates or forwards CloudEvents to its destination MUST produce

Reviewer 3: `SHOULD` more than `MUST`? Do we expect all sources, including `ContainerSource`, to fullfill this?

Author: As discussed during the WG, this is now a SHOULD with a sentence describing cases when this is not possible/desirable

or extend (respectively) the CloudEvents with the `knsource` extension, with

> value following the format `plural.group/name`. Off cluster sources MAY produce

Reviewer 2: Could you clarify what `plural.group/name` means? Without expansion I am not sure what is expected as the value.

the `knsource` extension.

> 

Reviewer 4: Explain here what is difference between `knsource` and Cloud Event `source` and `subject` attribute? Somehing like:

```
The use case addressed is to identify sources that generate or forwards events in context of Knative cluster where eventing is installed. The existing Cloud Event attributes `source` and `subject` may be already present in forwarded event and changing them would be surprising to events consumers.
```

Author: Added, thanks!

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

### CloudEvent `knsource` extension

Sources SHOULD produce CloudEvents with the `knsource` CloudEvent extension
identifying the originating [Source Custom Object](#source-custom-objects). It

> is highly-recommended to produce the `knsource` CloudEvent extension.

Reviewer 5: Isn't `SHOULD` read as "(highly) recommended"already ?

Author: yes, but you know, everyone can use a bit of extra motivation .... Anyway I'm happy to remove it.

Author: I removed.

> identifying the originating [Source Custom Object](#source-custom-objects). It
> is highly-recommended to produce the `knsource` CloudEvent extension.

Reviewer: ```suggestion
identifying the originating [Source Custom Object](#source-custom-objects). It
is highly-recommended to produce the `knsource` CloudEvent extension because without one, users might not be able to filter events coming from different custom objects that belong to the same Source CRD. See the Examples section below.
```
Just another suggestion.

Reviewer 2: That's what I also suggested, clarify the _why_ in the same sentence.

A correction though: objects do not "belong" to a CRD. If we use your phrasing, it should read "objects of a certain source kind".

Author: I feel this does not belong in the spec, which is meant to be concise and prescriptive. I would rather keep the use case example in a separate paragraph.

Reviewer: @lionelvillard fair enough, what do you think of this version of wording?

The `knsource` value SHOULD be of the form `<name>.<plural>.<group>`, where:

- `name` is the [Source Custom Object](#source-custom-objects) name
- `plural` is the [Source CRD](#source-crds)
  [plural](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)
- `group` is the [Source CRD](#source-crds)
  [API group](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)

Examples:

- `hello.pingsources.sources.knative.dev`

> - `rabbitmq-source.rabbitmq.sources.knative.dev`

Reviewer 5: Isn't that `...rabbitmqsources.sources.knative.dev`, instead of `rabbitmq.sources.knative.dev`?

See: https://github.com/knative-sandbox/eventing-rabbitmq/blob/master/config/source/300-rabbitmqsource.yaml#L28

Author: Good catch, thanks!

It is allowed to specify an alternative value in
[`spec.ceOverrides.extensions`](#duckspec), in which case the `knsource` value
can be of any shape.

When the `knsource` CloudEvent extension already exists in the CloudEvent being
forwarded by a Source, its value SHOULD be overridden as described above.

One use case addressed by the `knsource` CloudEvent extension is event filtering
based on [Source Custom Object](#source-custom-objects). For instance triggers
can be configured to receive events coming from a single
[Source Custom Object](#source-custom-objects).

> 

Reviewer 5: Do we have more than one use-case ?

Not really sure what this is trying to solve.

/assign @slinkydeveloper 
/assing @n3wscott

Author: Here the full context: https://github.com/knative/eventing/issues/1381

TL;TR: `One of the key motivations of the work around the origin of this issue is to support the FaaS scenario and make it possible to write side-effect free triggers.`.

Hopefully I'm not opening a can of worms by referring this long discussion. 😨

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

### CloudEvent `knsource` extension

Sources SHOULD produce CloudEvents with the `knsource` CloudEvent extension
identifying the originating [Source Custom Object](#source-custom-objects). It

> is highly-recommended producing the `knsource` CloudEvent extension.

Reviewer 2: ```suggestion
is highly-recommended to produce the `knsource` CloudEvent extension.
```

Maybe we should add a note why, e.g. _"So that subscribers can consume events originating from a specific source instance."_

It's kind of explained in the last paragraph of this section but both paragraphs feel a bit disconnected to me.

Author: I'll blame the stupid grammar checker. 

The reason I put the use case at the end is to make the spec as clear and concise as possible. Maybe explicitly stating it's a use case may help?

Reviewer 2: I think it would help to have it in the same sentence so the reader immediately understands _why_, otherwise the explanation comes much later.

cc @devguyio who suggested the same change pretty much.

The `knsource` value SHOULD be of the form `<plural>.<group>/<name>`, where:

- `plural` is the [Source CRD](#source-crds)
  [plural](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)
- `group` is the [Source CRD](#source-crds)
  [API group](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)
- `name` is the [Source Custom Object](#source-custom-objects) name

Examples:

- `pingsources.eventing.knative.dev/hello`

> - `rabbitmq.sources.knative.dev/rabbitmq-source`

Reviewer 2: 👍

Author: good!

It is allowed to specify an alternative value in
[`spec.ceOverrides.extensions`](#duckspec), in which case the `knsource` value
can be of any shape.

When produced, the `knsource` CloudEvent extension MUST appear in

> [status.ceAttributes](#duckstatus).

Reviewer 2: How? `CloudEventAttributes` contains only `Type` and `Source`.

Author: oh good catch! That's right, not possible. I removed this statement for now and I'll open an issue.

> The `knsource` CloudEvent extension already exist in the CloudEvent being
> forwarded by a Source, its value SHOULD be overridden as described above.

Reviewer 2: Did you forget the "If" or am I reading it wrong?

```suggestion
If the `knsource` CloudEvent extension already exists in the CloudEvent being
forwarded by a Source, its value SHOULD be overridden as described above.
```

Author: fixed

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

### CloudEvent `knsource` extension

Sources SHOULD produce CloudEvents with the `knsource` CloudEvent extension
identifying the originating [Source Custom Object](#source-custom-objects).

The `knsource` value SHOULD be of the form `<name>.<plural>.<group>`, where:

- `name` is the [Source Custom Object](#source-custom-objects) name
- `plural` is the [Source CRD](#source-crds)
  [plural](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)
- `group` is the [Source CRD](#source-crds)
  [API group](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)

Examples:

- `hello.pingsources.sources.knative.dev`

> - `rabbitmq-source.rabbitmqsources.sources.knative.dev`

Reviewer 2: That example sounds confusing to me because `name` repeats parts of both the `plural` and the `group`. IMO examples in a spec should be _extra careful_ about not confusing the reader in any manner.

```suggestion
- `myqueue.rabbitmqsources.sources.knative.dev`
```

Author: That's real: https://github.com/knative-sandbox/eventing-rabbitmq/blob/master/samples/source/event-source.yaml :-)

Changed.

It is allowed to specify an alternative value in
[`spec.ceOverrides.extensions`](#duckspec), in which case the `knsource` value
can be of any shape.

When the `knsource` CloudEvent extension already exists in the CloudEvent being

> forwarded by a Source, its value SHOULD be overridden as described above.

Reviewer 2: Just to avoid any misinterpretation such as "if the value is already set you must set the extension" (which is not the intended meaning at all). That's exactly how I read it, which means I won't be the only one probably.

```suggestion
forwarded by a Source, its value SHOULD be overridden when the value described
above is set.
```

One use case addressed by the `knsource` CloudEvent extension is event filtering

> based on [Source Custom Object](#source-custom-objects). For instance triggers
> can be configured to receive events coming from a single
> [Source Custom Object](#source-custom-objects).
> 

Reviewer 2: A lot of nits in this single sentence:
* commas
* do we need to capitalize "custom" and "object"?
* let's avoid mixing capitalization when referring to API objects (either all start with a capital letter, or none)
* I think the spec means "specific" rather than "single" here (as in "this one" versus "just one")
* that one is very subjective but the sentence sounded unfinished to me

```suggestion
based on [Source custom object](#source-custom-objects). For instance, Triggers
can be configured to receive events coming from a specific [Source custom
object](#source-custom-objects) by filtering events on this context attribute.
```

Author: See this thread about capitalization: https://github.com/knative/eventing/pull/3742#discussion_r462447985

Reviewer 2: Fair enough.

Regarding the last 2 points, do I understand correctly that you prefer not to include them?

Author: nope, this is good. Added.

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

### CloudEvent `knsource` extension

> Sources SHOULD produce CloudEvents with the `knsource` CloudEvent extension

Reviewer 6: ```suggestion
Sources MAY produce CloudEvents with the `knsource` CloudEvent extension
```

Author: I'm fine with that. @duglin ?

identifying the originating [Source Custom Object](#source-custom-objects).

> The `knsource` value SHOULD be of the form `<name>.<plural>.<group>`, where:

Reviewer 6: Ideally the adapters we need to use to bridge cloudevents are transparent and don't decorate the event. I am not sure what value it brings to add metadata on how the event occurrence was brought into the cluster. If that data is not understood or worthy of being part of `type`, `source` or `subject does it really have value to the routers and consumers of the event?

Author: > Ideally the adapters we need to use to bridge cloudevents are transparent and don't decorate the event

Agree, that's the main reason for the SHOULD.

> I am not sure what value it brings to add metadata on how the event occurrence was brought into the cluster.  If that data is not understood or worthy of being part of `type`, `source` or `subject` does it really have value to the routers and consumers of the event?

The originating CR ref cannot be part of of `type`, `source` or `subject` because their values are usually defined outside of Knative. There are exceptions though (eg. PingSource). It is useful in the context of Knative, and Knative only.

Reviewer 6: I think that is so source specific it makes no sense to add it to the spec.

Reviewer: @n3wscott I'm not sure what do you mean by "source specific"? what about the use cases described above? it'd be great if you check them and let us know how else can a knative developer specify a filter when `type`, `source` and `subject` are prepopulated or too broad ? IIRC every knative source can conform very easily to `knsource` cause every source can fill the `<name>.<plural>.<group>` combination, so I'm not sure why is it too source specific?

> The `knsource` value SHOULD be of the form `<name>.<plural>.<group>`, where:

Reviewer 6: ```suggestion
The `knsource` value MUST be of the form `<name>.<plural>.<group>`, where:
```

If there is no known way to understand this value, it likely has no meaning. 

I would recommend inventing a uri for this instead of a magic string, like `k8s://<plural>.<group>/<namespace>/<name>/<uid>` or something like that.

But l think likely if you care this much about this level of detail of _who_ you might also care about the pod that is doing the work.

Author: The reason it is SHOULD and not MUST is the value can be overridden. What about `The default knsource value MUST ...`?

We can invent a magic uri if you want, it does not make a difference IMO. Independently of the actual format, I think we agree that `plural`, `group` and `name` should be in there. 

`uid` is a dynamic value making the deployment of Triggers much more complicated. It's also much easier for the user to understand where the events are coming from if the value only contains elements (name, plurals, group) that are specified. Maybe `<plural>` should even be `<kind>` (lower-case).

As for `namespace`, it's a bit more complicated. I'd like to be able to specify filters that are namespace-unaware. I can't think of a better solution than to add another extension, `knsourcens`, at least until we have a better filtering language.

> - `name` is the [Source Custom Object](#source-custom-objects) name

Reviewer 6: name is not interesting without namespace is my guess.

Author: added ns.

> - `plural` is the [Source CRD](#source-crds)

Reviewer 6: this is just called `resource`.

Author: I added the namespace as part of the value.

@n3wscott I would rather stick with `plural` as that's how it is called in CRDs.

@rhuss this extension attribute is about identifying the event source object that has been used to subscribe to events. For Kamelet, if you want to filter based on the source you connect to, either use the event type or event source. Does this make sense?

Reviewer 6: I would drop all of these and use resource `<uid>-<generation>`

  [plural](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)
- `group` is the [Source CRD](#source-crds)

>   [API group](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)

Reviewer 7: I am guessing this is a step to enable consuming events from a specific source, are these with the source field enough to identify a source? eg same name but in different namespace or even cluster.

Author: `with the source field`: what do you mean?

Examples:

> - `hello.pingsources.sources.knative.dev`

Reviewer 6: `hello` is not  enough to tell me which pingsource it is coming from.

Author: Why? missing namespace?

Reviewer 6: it is missing namespace, and instance id if you want to be every direct with which resource made this event.

- `myqueue.rabbitmqsources.sources.knative.dev`

It is allowed to specify an alternative value in
[`spec.ceOverrides.extensions`](#duckspec), in which case the `knsource` value
can be of any shape.

> When the `knsource` CloudEvent extension already exists in the CloudEvent being

Reviewer 6: This is very confusing.

Author: What do you mean? Not sure how trace headers can help.

Reviewer 6: it seems like you want to filter based on the path the event took to the broker and if we could record that data and link the trace span with a source instance, it might get what you want.

Reviewer 8: I think one of the use-cases is the original [faas scenario](https://docs.google.com/document/d/1DpiSL2dUcYS2n7yXOIG5LJwyIC1lY9q_W8-56U1SvKM/edit?hl=en) where a trigger and the source are instantiated and chained together in one go.

SHOULD send events to its
[Destination](https://pkg.go.dev/github.com/knative/pkg/apis/duck/v1?tab=doc#Destination).

### CloudEvent `subscription` extension

Sources SHOULD produce CloudEvents with the `subscription` CloudEvent extension
identifying the originating [Source Custom Object](#source-custom-objects).

> The `subscription` value SHOULD be of the form `<name>.<namespace>.<plural>.<group>`, where:

Reviewer 6: this is still wrong. the name of the resource is not the subscription, it is the uid of the resource and maybe a generation for that resource.

Reviewer 9: Can we guarantee that from an end-user perspective the source (cr) will always be created before and in a distinct step from the creation of a trigger/filter?  If so, then uid might be ok - but some user-friendly info would sure be a lot nicer UX. I'm not sure why we're so concerned about using the name here when the rest of Kube typically does refs by name and not uid.

Not sure about generation though since I wouldn't want an update to a source (cr) to cause everyone to recreate their filters and lose events in the meantime.
